### ISSUE TITLE
UnicodeDecodeError when creating new group

### ISSUE THREAD

**Issue by matclab** [web2py/pydal#403]
_Seen with pydal v16.03._

In web2py, when a new user is registering, a new group is created with a localised group name. for example, in french : 'Groupe unique affecté à l'utilisateur 6369470859706368'.

Some excerpt of a debugging session are showned here after:
```
/home/clabaut/web/web2py-lost.re/gluon/packages/dal/pydal/objects.py(726)insert()
-> ret = self._db._adapter.insert(self, self._listify(fields))
(Pdb) p fields
{'role': 'user_6369470859706368', 'description': <lazyT 'Group uniquely assigned to user %(id)s'>}
…
/home/clabaut/web/web2py-lost.re/gluon/packages/dal/pydal/adapters/base.py(1937)represent()
-> obj = to_unicode(obj)
(Pdb) p obj
<lazyT 'Group uniquely assigned to user %(id)s'>
(Pdb) unicode(obj)
(Pdb) str(obj)
"Groupe unique affect\xc3\xa9 \xc3\xa0 l'utilisateur 6369470859706368"
```

This seems similar to a web2py bug corrected in [web2py PR 1163](https://github.com/web2py/web2py/pull/1163).

Using `str` instead of `unicode` everytime or upon `lazyT` objects may be the solution.

**Comment by matclab** [comment]
We should perhaps call `obj.decode` for any object that has a callable `decode` argument (it would apply to `bytes` and `lazyT`.

Any advice appreciated…

**Comment by matclab** [comment]
The following patch might do the trick:
```diff
diff --git a/pydal/_compat.py b/pydal/_compat.py
index 97e9194..d0efd64 100644
--- a/pydal/_compat.py
+++ b/pydal/_compat.py
@@ -86,9 +86,10 @@ def with_metaclass(meta, *bases):
 def to_unicode(obj, charset='utf-8', errors='strict'):
     if obj is None:
         return None
-    if not isinstance(obj, bytes):
+    try:
+        return obj.decode(charset, errors)
+    except AttributeError:
         return text_type(obj)
-    return obj.decode(charset, errors)

```

**Comment by gi0baro** [comment]
I don't agree on this. `lazyT` is a web2py object, the correct methods implementation for Unicode handling should be done in web2py. Patching pydal's compat module like this means only slowing down the whole source code every time `to_unicode` is invoked.

**Comment by matclab** [comment]
@gi0baro, I agree that we shall not do specific things for web2py objects.

However, it looks not so bad to me to generalize the code used for `bytes`, for all objects that have a `decode` method (as proposed in my last comment). What do you think ?
(`lazyT` allows for correct unicode handling by the use of such method).

**Comment by gi0baro** [comment]
Still, the `to_unicode` method was written to work principally with strings, and eventually to "unicodify" the string representation of non-string objects. I keep thinking if we need a "decode" method to generally decode objects to unicode, then let's write another method, rather than patching something is working good for the purpose it was written for.

**Comment by matclab** [comment]
I'm totally new to the pydal codebase, so I have no good advice on the
subject. If you think the problem belongs to web2py I can open a issue over
there. But for the moment I've a non working code.
If there is no change on pydal side, I'll try to see if I can find a patch
on the web2py side and keep you inform here.
Thanks for your help...

Le mar. 9 août 2016 20:00, Giovanni Barillari <notifications@github.com> a
écrit :

**Comment by gi0baro** [comment]
@matclab I'm sorry I was travelling and I had no chance to reply to this sooner. I think the best option right now is to open up an issue on web2py repository, since this is related to the translator.
